README
Freeware Advanced Audio Coder/Decoder
http://www.audiocoding.com/


__________
COPYRIGHTS

FAAD is based on the ISO MPEG-4 reference code. For this base code the
following license applies:


**********************************************************************
This software module was originally developed by

FirstName LastName (CompanyName)

and edited by

FirstName LastName (CompanyName)
FirstName LastName (CompanyName)

in the course of development of the MPEG-2 NBC/MPEG-4 Audio standard
ISO/IEC 13818-7, 14496-1,2 and 3. This software module is an
implementation of a part of one or more MPEG-2 NBC/MPEG-4 Audio tools
as specified by the MPEG-2 NBC/MPEG-4 Audio standard. ISO/IEC gives
users of the MPEG-2 NBC/MPEG-4 Audio standards free license to this
software module or modifications thereof for use in hardware or
software products claiming conformance to the MPEG-2 NBC/ MPEG-4 Audio
standards. Those intending to use this software module in hardware or
software products are advised that this use may infringe existing
patents. The original developer of this software module and his/her
company, the subsequent editors and their companies, and ISO/IEC have
no liability for use of this software module or modifications thereof
in an implementation. Copyright is not released for non MPEG-2
NBC/MPEG-4 Audio conforming products. The original developer retains
full right to use the code for his/her own purpose, assign or donate
the code to a third party and to inhibit third party from using the
code for non MPEG-2 NBC/MPEG-4 Audio conforming products. This
copyright notice must be included in all copies or derivative works.

Copyright (c) 1997.
**********************************************************************


For the changes made for the FAAC project the GNU Library General
Public License (LGPL), version 2 1991 applies. For the changes the
following statement applies:

**********************************************************************
FAAD - Freeware Advanced Audio Decoder
Copyright (C) 2001 M. Bakker

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
**********************************************************************



Please note that the use of this software may require the payment of
patent royalties. You need to consider this issue before you start
building derivative works. We are not warranting or indemnifying you in
any way for patent royalities! YOU ARE SOLELY RESPONSIBLE FOR YOUR OWN
ACTIONS!


______
PEOPLE

Currently maintained by
M. Bakker (menno@audiocoding.com).

Other contributors
Tony Lenox (len0x@mail.ru)
RageOMatic (rageamp@realedge.com)
thebard (wmilas@rarcoa.com)
Ivan Dimkovic (dim@psytel-research.co.yu)

(If you feel your name should be here please mail me, I
can't remember everyone)


_________
COMPILING

Compiling under Unix:
Taken from a message by ed at the audiocoding.com forum. Thanks!

the linux tools are munted by the CR/LFs all over the place ... to make
then happy once again, all lines need to be terminated by LF alone

how to sort this problem ?, well stick the following script into a file,
unzip faad, run the script on the faad directory (it'll recursively fix
all files), then continue with sh bootstrap ....


#!/bin/bash

if [ -d $1 ]; then
for i in $1/*; do
./$0 $i
done
else
mv $1 tmp
cat tmp | tr -d \\r > $1
rm tmp
fi


ChangeLog
2001-08-31  menno
		Little bugfix in intialising an array

2001-08-18  rageomatic
		speed and streaming updates

2001-07-10  menno
		Added Sonique plugin (written by Nic)

2001-07-09  menno
		Added project file for compiling faad as a DLL

2001-07-03  rageomatic
		Added OK and Cancel buttons

2001-07-02  menno
		Fix for using faac and faad in the same program

2001-07-01  rageomatic
		Small bug with the variable bitrate check box

		Added buffering options

2001-06-28  rageomatic
		Updated configuration UI. A few new features.

2001-06-27  menno
		Basically all ID3 text frames are read now

		Unix compile fix

		Redone ID3v2 reading
		Every field will be shown now

2001-06-26  menno
		Compile fix under Unix

2001-06-25  menno
		Fixed assertion failure

		Compilation help

2001-06-22  menno
		Some compatibility issues are removed .
		A bug removed.
		Better Error messages
		Some comments in the source.
		In-Place correction possible.
		Needs ANSI + POSIX 1.0.
		Thanks to Frank Klemm

2001-06-12  menno
		bugfix

		new option parsing

2001-06-11  menno
		removed assert()

2001-06-10  menno
		Small cleanup

		Fixed memory leaking
		Output to stderr, so that writing to stdio still gives some info on screen

		Now showing percentage while decoding

2001-06-09  menno
		Added util to remove PsyTEL tags

2001-06-08  menno
		File renamed

		Added license files for third party code

		Fixed linux compilation

2001-06-07  menno
		Made changes to the configuration structure.
		Now the default samplerate and object type can be set for use with headerless AAC files

2001-06-06  menno
		Small cleanup

		(I hope I have) fixed MPEG4 compatibility once and for all ;-)

2001-06-01  menno
		Fixed seeking for MPEG4 files

2001-05-30  menno
		Changed license from GPL to LPGL

		Added CR

2001-05-29  menno
		Small changes to the info box
		Disabled seeking for MPEG4 AAC files temporarily

		Fixed linux compilation

2001-05-28  menno
		Inclusion of id3lib

		Bug fixes and some changes for nicer compilation (Thanks to
		Bill May from MPEG4IP)

		Support for reading of some ID3v2 tag fields (not all yet)

		Changes for ISO MPEG4 compatibility (sorry :(( )
		Old MPEG4 files with ADTS header will not be playable anymore

		Fixed installer so that id3lib.dll gets copied to the system directory

2001-05-20  menno
		Small code cleanup

		New file info dialog, just got to implement ID3v2 tags now.

2001-05-17  menno
		Added decoder library documentation

2001-05-14  menno
		Added MP4 file library
		and an utilitie to create MP4 files from aac files and one for extracting audio from a MP4 file

2001-05-11  oxygene2000
		 New makesystem for unix-like systems based on autoconf/-make
		and libtool

		- didn't compile on !win32 because of missing ;

2001-05-09  menno
		More stable winamp plugin, stopping a file should work correctly now
		Fixed NSIS install script to avoid warnings

2001-05-02  menno
		Removed unused file stdinc.h

		Added correct copyright statements
		Restored thread safety

2001-04-29  rageomatic
		Speed, stability and streaming improvements.

		Slight speed improvements.

2001-04-27  menno
		Made most buffers dynamically allocated

		Added compile option for Intel signal processing library

2001-04-26  menno
		Fixes for streaming
		Windows API used instead of standard library

2001-04-25  menno
		Changed command line option -o, so that it supports directories
		with spaces

2001-04-25  rageomatic
		New streaming design, uses Winsock instead of WinInet now.

2001-04-24  menno
		Multichannel decoding support

2001-04-23  menno
		Better file buffering for winamp plugin
		Better windows API support

2001-04-22  menno
		Added backward prediction
		FAAD can now decode MAIN profile AAC files

		Made backward prediction thread safe too

		Added util.c file for helper functions

		Bug fix

2001-04-21  menno
		Changes to library interface
		-better error handling
		-better multichannel support

		More multichannel fixes

		Better DSE and FIL element support

		Last change to library interface

2001-04-20  menno
		Fixed ADIF decoding

2001-04-19  menno
		FFT code is thread safe now too

		Made code thread safe.
		Only FFT code has to be done now.

		Changes in MPEG4 file format, mainly comments for clarification

2001-04-16  menno
		Higher thread priority for plugin
		Makefile fixed

2001-04-11  menno
		Fixed MPEG4 object type problem
		Use new adtsfix util to make old AAC file playable again

		Small updates

2001-04-10  menno
		- major code cleanup
		- better error handling
		- tables used for KBD window when WIN_TABLE defined
		- Somewhat better thread safety (not nearly thread safe, but a lot of data has moved)

		Added adts_error_check

		small fix

		Better error recovery for ADTS files

		Fixed another MPEG4/2 issue TNS order

		Removed http streaming for now, somehow it didn't work anymore
		fix will come later

2001-04-06  menno
		Bugfixes

		Faster bit reading routines

2001-04-05  menno
		Fixed Mpeg2 bug

		ADTSFIX utility for fixing old AAC files

2001-04-04  menno
		Another filterbank speedup

		Speedup in roundof

		Speedup in filterbank

2001-04-03  menno
		Big changes to library interface

		(more) correct decoding of mpeg2-aac

2001-03-30  menno
		Bugfix

2001-03-29  menno
		Speedup (thanks to Ivan)

		Simple file check so that plugin doesn't crash if file doesn't exist

		Major bugfix in bits.c, no more weird output :)

2001-03-22  menno
		Clipping problem fixed

2001-03-20  menno
		Bugfixes

2001-03-19  menno
		Bugfixes, mainly in channel configuration and file properties
		Correct decoding of files other that 44.1kHz and 2 channels now

		Moved Winamp plugin to separate directory

2001-03-18  menno

winamp/aacinfo.c
/*
 * FAAD - Freeware Advanced Audio Decoder
 * Copyright (C) 2001 Menno Bakker
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * $Id: aacinfo.c,v 1.22 2001/09/24 07:31:58 menno Exp $
 */

#define WIN32_MEAN_AND_LEAN
#include <windows.h>
#include "filestream.h"
#include "aacinfo.h"

#define ADIF_MAX_SIZE 30 /* Should be enough */
#define ADTS_MAX_SIZE 10 /* Should be enough */

const int sample_rates[] = {96000,88200,64000,48000,44100,32000,24000,22050,16000,12000,11025,8000};

static int read_ADIF_header(FILE_STREAM *file, faadAACInfo *info)
{
    int bitstream;
    unsigned char buffer[ADIF_MAX_SIZE];
    int skip_size = 0;
    int sf_idx;

    /* Get ADIF header data */
    info->headertype = 1;

    if(read_buffer_filestream(file, buffer, ADIF_MAX_SIZE) < 0)
		return -1;

    /* copyright string */
    if(buffer[4] & 128)
        skip_size += 9; /* skip 9 bytes */

    bitstream = buffer[4 + skip_size] & 16;
    info->bitrate = ((unsigned int)(buffer[4 + skip_size] & 0x0F)<<19)|
        ((unsigned int)buffer[5 + skip_size]<<11)|
        ((unsigned int)buffer[6 + skip_size]<<3)|
        ((unsigned int)buffer[7 + skip_size] & 0xE0);

    if (bitstream == 0) {
        info->object_type = ((buffer[9 + skip_size]&0x01)<<1)|((buffer[10 + skip_size]&0x80)>>7);
        sf_idx = (buffer[10 + skip_size]&0x78)>>3;
        info->channels = ((buffer[10 + skip_size]&0x07)<<1)|((buffer[11 + skip_size]&0x80)>>7);
    } else {
        info->object_type = (buffer[7 + skip_size] & 0x18)>>3;
        sf_idx = ((buffer[7 + skip_size] & 0x07)<<1)|((buffer[8 + skip_size] & 0x80)>>7);
        info->channels = (buffer[8 + skip_size]&0x78)>>3;
    }
    info->sampling_rate = sample_rates[sf_idx];

    return 0;
}

static int read_ADTS_header(FILE_STREAM *file, faadAACInfo *info, int *seek_table, int tagsize)
{
    /* Get ADTS header data */
    unsigned char buffer[ADTS_MAX_SIZE];
    int frames, t_framelength = 0, frame_length, sr_idx, ID;
    int second = 0, pos;
	int i;
    float frames_per_sec = 0;
    unsigned long bytes;

    info->headertype = 2;

    /* Read all frames to ensure correct time and bitrate */
    for(frames=0; /* */; frames++)
    {
		/* If streaming, only go until we hit 5 seconds worth */
		 if(file->http)
		 {
			 if(frames >= 43 * 5)
			 {
				break;
			 }
		 }

        /* 12 bit SYNCWORD */
        bytes = read_buffer_filestream(file, buffer, ADTS_MAX_SIZE);

        if(bytes != ADTS_MAX_SIZE)
        {
            /* Bail out if no syncword found */
            break;
        }

		/* check syncword */
        if (!((buffer[0] == 0xFF)&&((buffer[1] & 0xF6) == 0xF0)))
            break;

        pos = tell_filestream(file) - ADTS_MAX_SIZE;

        if(!frames)
        {
            /* fixed ADTS header is the same for every frame, so we read it only once */
            /* Syncword found, proceed to read in the fixed ADTS header */
            ID = buffer[1] & 0x08;
            info->object_type = (buffer[2]&0xC0)>>6;
            sr_idx = (buffer[2]&0x3C)>>2;
            info->channels = ((buffer[2]&0x01)<<2)|((buffer[3]&0xC0)>>6);

            frames_per_sec = sample_rates[sr_idx] / 1024.f;
        }

        /* ...and the variable ADTS header */
        if (ID == 0) {
            info->version = 4;
            frame_length = (((unsigned int)buffer[4]) << 5) |
                ((unsigned int)buffer[5] >> 3);
        } else { /* MPEG-2 */
            info->version = 2;
            frame_length = ((((unsigned int)buffer[3] & 0x3)) << 11)
                | (((unsigned int)buffer[4]) << 3) | (buffer[5] >> 5);
        }

        t_framelength += frame_length;

		if(!file->http)
		{
			if (frames > second*frames_per_sec)
			{
				seek_table[second] = pos;
				second++;
			}
		}

		/* NOTE: While simply skipping ahead by reading may seem to be more work than seeking,
			it is actually much faster, and keeps compatibility with streaming */
		for(i=0; i < frame_length - ADTS_MAX_SIZE; i++)
			if(read_byte_filestream(file) < 0)
				break;
    }

    info->sampling_rate = sample_rates[sr_idx];
    info->bitrate = (int)(((t_framelength / frames) * (info->sampling_rate/1024.0)) +0.5)*8;

	if(file->http)
	{
		/* Since we only use 5 seconds of aac data to get a rough bitrate, we must use a different
		   method of calculating the overall length */
		if(filelength_filestream(file))
		{
			info->length = (int)((filelength_filestream(file)/(((info->bitrate*8)/1024)*16))*1000);
		}
		else
		{
			/* Since the server didnt tell us how long the file is, 
			   we have no way of determining length */
			info->length = 0;
		}
	}
	else
	{
		info->length = (int)((float)(frames/frames_per_sec))*1000;
	}

    return 0;
}

int get_AAC_format(char *filename, faadAACInfo *info, int *seek_table)
{
    unsigned long tagsize;
    FILE_STREAM *file;
	char buffer[10];
    unsigned long file_len;
    unsigned char adxx_id[5];
    unsigned long tmp;

    file = open_filestream(filename);

    if(file == NULL)
        return -1;

    file_len = filelength_filestream(file);

    /* Skip the tag, if it's there */
    tmp = read_buffer_filestream(file, buffer, 10);

    if (StringComp(buffer, "ID3", 3) == 0)
	{
		int i;

        /* high bit is not used */
        tagsize = (buffer[6] << 21) | (buffer[7] << 14) |
            (buffer[8] <<  7) | (buffer[9] <<  0);

        for(i=0; i < tagsize; i++)
			if(read_byte_filestream(file) < 0)
				return -1;

        tagsize += 10;
    }
	else
	{
        tagsize = 0;

		/* Simple hack to reset to the beginning */
		file->buffer_offset = 0;
		file->file_offset = 0;
    }

	if(file_len)
	    file_len -= tagsize;

    tmp = read_buffer_filestream(file, adxx_id, 2);
    //seek_filestream(file, tagsize, FILE_BEGIN);

    adxx_id[5-1] = 0;
    info->length = 0;

	/* Determine the header type of the file, check the first two bytes */
    if(StringComp(adxx_id, "AD", 2) == 0)
    {
		/* We think its an ADIF header, but check the rest just to make sure */
		tmp = read_buffer_filestream(file, adxx_id + 2, 2);
		
		if(StringComp(adxx_id, "ADIF", 4) == 0)
		{
			read_ADIF_header(file, info);
		}
    }
    else
    {
		/* No ADIF, check for ADTS header */
        if ((adxx_id[0] == 0xFF)&&((adxx_id[1] & 0xF6) == 0xF0))
        {
			/* ADTS  header located */
			/* Since this routine must work for streams, we can't use the seek function to go backwards, thus 
			   we have to use a quick hack as seen below to go back where we need to. */
			
			if(file->buffer_offset >= 2)
			{
				// simple seeking hack, though not really safe, the probability of it causing a problem is low.
				file->buffer_offset -= 2;
			}

            read_ADTS_header(file, info, seek_table, tagsize);
        }
        else
        {
            /* Unknown/headerless AAC file, assume format: */
            info->version = 2;
            info->bitrate = 128000;
            info->sampling_rate = 44100;
            info->channels = 2;
            info->headertype = 0;
            info->object_type = 1;
        }
    }

    close_filestream(file);

    return 0;
}

int StringComp(char const *str1, char const *str2, unsigned long len)
{
    signed int c1 = 0, c2 = 0;

    while (len--) {
        c1 = *str1++;
        c2 = *str2++;

        if (c1 == 0 || c1 != c2)
            break;
    }

    return c1 - c2;
}
